### DISCUSSION TITLE
Is it possible to skip mapping configuration for some env vars?

### DISCUSSION THREAD

**Discussion by ppatierno** [quarkusio/quarkus#33557] [Q&A] [answered]
Hi all,
I have the following `KafkaConfig` class in our Strimzi HTTP bridge (based on Quarkus) https://github.com/ppatierno/strimzi-kafka-bridge/blob/road-to-quarkus/src/main/java/io/strimzi/kafka/bridge/config/KafkaConfig.java

With this class we are able to get configuration from an `application.properties` file like this https://github.com/ppatierno/strimzi-kafka-bridge/blob/road-to-quarkus/config/application.properties

You can see we are mapping `kafka.producer.`, `kafka.consumer.` and `kafka.admin.` prefixed properties to corresponding maps into the `KafkaConfig`. Also everything under just `kafka.` (which doesn't match above prefixes) goes to `common` map.

The problem I am facing is that the environment where this application runs sets some env vars (used for a different purpose) starting with `KAFKA_BRIDGE_` so the Quarkus configuration is trying to map them into the `KafkaConfig.common` map ... which doesn't make any sense for our application.
Furthermore, some of these env vars could be empty and the Quarkus application doesn't start at all because of error like this (taking `KAFKA_BRIDGE_PRODUCER_CONFIG` env var as an example):

```shell
ERROR: Failed to start application (with profile [prod])
java.util.NoSuchElementException: SRCFG00040: The config property kafka.bridge.producer.config is defined as the empty String ("") which the following Converter considered to be null: io.smallrye.config.Converters$BuiltInConverter
```

So my question is ... is there any way to avoid some prefixed env vars to be mapped (i.e. all starting with `KAFKA_BRIDGE_`) or disabling the env var config source at all (because we don't need that)?
I tried to skip `KAFKA_BRIDGE_` prefixed env vars by using a config relocator returning just `null` when it's queried for properties this way:

```shell
@Override
public ConfigValue getValue(ConfigSourceInterceptorContext context, String name) {
...
...
   if (name.startsWith("kafka.bridge.")) {
            return null;
   } 
}
```

I also tried a really BAD hack by adding trash "bucket" (actually a Map) into the `KafkaConfig` class to get all these env vars, but it doesn't work as well.
Any suggestions? (other then renaming the env vars in our environment to avoid them starting with `KAFKA_BRIDGE_` prefix ;-))

**Comment by radcortez** [comment, marked as answer]
Currently, there is no way to do it.

When mapping properties, the Config system doesn't care from where the config property is coming from.

I've suggested the interceptor and remove the properties. Still, unfortunately, that doesn't work because we always apply another interceptor on top (no matter the priority), to convert environment variables to their dotted format.

The bucket solution should work, but remember to quote the `kafka.bridge` in the interceptor. Also, you need to add a converter to handle empty values, because we always expect to have a value so you can do something like:

```java
    Map<String, @WithConverter(IgnoreConverter.class) String> bridge();

    class IgnoreConverter implements Converter<String> {
        @Override
        public String convert(final String s) throws IllegalArgumentException, NullPointerException {
            return "ignored";
        }
    }
```

**Reply by ppatierno** [reply]
Thanks Roberto, even if I don't think this is a really great way to fix this :-(
Would it make sense to fill an issue about this use case, allowing something like an "exclusion" support where you can specify what you don't want to be mapped during configuration process?

**Reply by radcortez** [reply]
That is already supported at the builder level, but the API is unavailable for the user (at least in Quarkus). I may think of a way to expose it, but it will take some time because it involves changes in both SmallRye Config and Quarkus. Unfortunately, I also have a few other priorities.

**Reply by ppatierno** [reply]
It's not a higher priority thing. I was suggesting to open an issue to track it as an enhancement which seems to make sense.

**Reply by radcortez** [reply]
Sure, go ahead. Thank you!

**Reply by ppatierno** [reply]
FYI https://github.com/quarkusio/quarkus/issues/33592, Thanks!
